issue_1/test/word_registry/TestWordRegistry.ts

```diff
@@ -0,0 +1,82 @@
+import hre from "hardhat";
+import { Artifact } from "hardhat/types";
+import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/dist/src/signer-with-address";
+
+import { expect } from "chai";
+import { WordRegistry } from "../../typechain/WordRegistry";
+import { Signers } from "../types";
+
+const { deployContract } = hre.waffle;
+
+const NULL_ADDRESS = '0x0000000000000000000000000000000000000000';
+
+describe("Unit tests", function () {
+  before(async function () {
+    this.signers = {} as Signers;
+
+    const signers: SignerWithAddress[] = await hre.ethers.getSigners();
+    this.signers.admin = signers[0];
+    this.signers.user1 = signers[1];
+    this.signers.user2 = signers[2];
+    console.log("admin: ", this.signers.admin.address);
+  });
+
+  describe("WordRegistry", function () {
+    beforeEach(async function () {
+      const artifact: Artifact = await hre.artifacts.readArtifact("WordRegistry");
+      this.wr = <WordRegistry>await deployContract(this.signers.admin, artifact, []);
+    });
+
+    it("should allow users to reserve words", async function () {
+      const word = "sss";
+      await expect(this.wr.connect(this.signers.user1).register(word))
+        .to.emit(this.wr, "WordRegistered").withArgs(word, this.signers.user1.address);
+    });
+
+    it("should preserve ownership of words", async function () {
+      const word = "sss";
+      expect(await this.wr.connect(this.signers.user1).wordToOwner(word))
+        .to.equal(NULL_ADDRESS);
+
+      await this.wr.connect(this.signers.user1).register(word);
+
+      expect(await this.wr.connect(this.signers.user1).wordToOwner(word))
+        .to.equal(this.signers.user1.address);
+    });
+
+    it("should allow users to unregister words", async function () {
+      const word = "sss";
+      await this.wr.connect(this.signers.user1).register(word);
```

Comment:
You can nest describe blocks (copied the code from a different example, but you'll be able to follow):
```
  describe("Registry", async () => {
    beforeEach(async function () {
      const registryArtifact: Artifact = await hre.artifacts.readArtifact("Registry");
      this.registry = <Registry>await deployContract(this.signers.admin, registryArtifact);
    });

    it("User should be able to claim a name", async function () {
      await expect(this.registry.connect(this.signers.admin).claimName("this")).to.emit(this.registry, "NameClaimed");
      expect(await this.registry.nameOwnership("this")).to.equal(await this.signers.admin.getAddress());
    });

    describe("with a claimed name", async () => {
      beforeEach(async function () {
        await this.registry.connect(this.signers.admin).claimName("this")
      });

      it("Users should be able to release a name", async function () { // The name has already been claimed!
        ...
      });

      it("Users should be able to claim a second name", async function () {
       ...
      });
   });
```
That way you follow a pattern:
 - You are at state 0
 - Test all the things that should revert
 - Test one thing that should work
 - Nest a describe, put the thing that works in the beforeEach, you are at state 1.

You will end with a kind of state tree, testing all the possible reverts, all the possible states, and all the possible successful functions.

Reply:
makes sense!